lucille-core/src/main/java/com/kmwllc/lucille/core/RunnerManager.java

```diff
@@ -1,80 +1,212 @@
+   * run with the given runId is already in progress.
+   *
+   * @param runId the unique ID for the lucille run
+   * @return RunDetails
+   * @throws RunnerManagerException
    */
-  public void waitForRunCompletion() throws ExecutionException, InterruptedException {
-    future.get();
+  public synchronized RunDetails run(String runId) throws RunnerManagerException {
+    Config config = ConfigFactory.load();
+    String configId = createConfig(config);
+    return runWithConfig(runId, configId);
   }
 
-  private CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+  /**
+   * Run with config and an internal runId will be generated
+   * 
+   * @param config
+   * @return RunDetails
+   * @throws RunnerManagerException
+   */
+  protected synchronized RunDetails runWithConfig(Config config) throws RunnerManagerException {
+    String configId = createConfig(config);
+    return runWithConfig(Runner.generateRunId(), configId);
+  }
 
   /**
-   * Main entrypoint for kicking off lucille runs. This method spawns a new thread for lucille to run in, but will not start a new
-   * instance of lucille until the previous one terminates.
+   * Method to create a lucille run with a custom configuration.
    *
-   * @return boolean representing whether the lucille run was initiated or not. This will return false if and only if the lucille
-   * run was skipped due to the previous run still existing.
+   * @param runId the unique ID for the lucille run
+   * @param config the configuration to use for the run
+   * @return true if the run was started successfully; false if a run with the given ID is already
+   *         in progress
    */
-  synchronized public boolean run() {
-    return runWithConfig(ConfigFactory.load());
+  public synchronized String createConfig(Config config) {
+    String configId = UUID.randomUUID().toString();
+    configMap.put(configId, config);
+    return configId;
   }
 
   /**
-   * Internal abstraction used to support testing.
+   * Method to start a lucille run with a custom configuration.
+   *
+   * @param runId the unique ID for the lucille run
+   * @param config the configuration to use for the run
+   * @return RunDetails
+   * @throws RunnerManagerException
    */
-  synchronized protected boolean runWithConfig(Config config) {
-    if (isRunning()) {
-      log.warn("Skipping new run; previous lucille run is still in progress.");
-      return false;
+  public synchronized RunDetails runWithConfig(String runId, String configId)
+      throws RunnerManagerException {
+
+    if (runId == null) {
+      throw new RunnerManagerException("runId cannot be null");
+    }
+    
+    if (runDetailsMap.containsKey(runId)) {
+      throw new RunnerManagerException("Run with runId " + runId + " already defined - cannot use this runId again");
+    }
+
+    final RunDetails runDetails = new RunDetails(runId, configId);
+    runDetailsMap.put(runId, runDetails);
+
+    if (isRunning(runId)) {
```

Comment:
you are throwing an exception above if `runDetailsMap.containsKey(runId)`
assuming that the runId is NOT in the map and you haven't thrown an exception, how could isRunning(runId) ever be true here, a few lines later?

Reply:
agreed,
isRunning(runId) was left over from previous lucille logic
removed